arcgis-ios-sdk-samples/Layers/List KML contents/ListKMLContentsSceneViewController.swift

```diff
@@ -0,0 +1,188 @@
+        let elevationSource = AGSArcGISTiledElevationSource(url: elevationSourceURL)
+        scene.baseSurface?.elevationSources.append(elevationSource)
+        
+        return scene
+    }()
+    
+    var kmlDataset: AGSKMLDataset? {
+        didSet {
+            
+            // clear any existing layers
+            scene.operationalLayers.removeAllObjects()
+            
+            guard let kmlDataset = kmlDataset else {
+                return
+            }
+            
+            // create a layer to display the dataset
+            let kmlLayer = AGSKMLLayer(kmlDataset: kmlDataset)
+            // add the kml layer to the scene
+            scene.operationalLayers.add(kmlLayer)
+        }
+    }
+    var node: AGSKMLNode? {
+        didSet {
+            title = node?.name
+            
+            if let node = node {
+                
+                // set the viewpoint based on the new node
+                setSceneViewpoint(for: node)
+            }
+        }
+    }
+    
+    //MARK: - Viewpoint
+    
+    /// Sets the viewpoint of the scene based on the node, or hides the node
+    /// if a viewpoint cannot be determined.
+    private func setSceneViewpoint(for node:AGSKMLNode){
+        
+        if !isViewLoaded {
+            // load the view before setting a viewpoint
+            loadView()
+        }
+        
+        if let nodeViewpoint = viewpoint(for: node),
+            !nodeViewpoint.targetGeometry.isEmpty {
+            
+            // reveal the view in case it was previously hidden
+            sceneView.isHidden = false
+            
+            // set the viewpoint for the node
+            sceneView.setViewpoint(nodeViewpoint)
+        }
+        else {
+            // this node has no viewpoint so hide the scene view, showing the info text
+            sceneView.isHidden = true
+        }
+    }
+    
+    /// Returns the elevation of the scene's surface corresponding to the point's x/y.
+    private func sceneSurfaceElevation(for point: AGSPoint) -> Double? {
+        guard let surface = scene.baseSurface else {
+            return nil
+        }
+        
+        var surfaceElevation: Double? = nil
+        let group = DispatchGroup()
+        group.enter()
+        // we want to return the elevation synchronously, so run the task in the background and wait
+        DispatchQueue.global(qos: .userInteractive).async {
+            surface.elevation(for: point, completion: { (elevation, error) in
+                if error == nil{
+                    surfaceElevation = elevation
+                }
+                group.leave()
+            })
+        }
+        // wait, but not longer than three seconds
+        _ = group.wait(timeout: .now() + 3)
+        return surfaceElevation
+    }
+    
+    /// Returns the viewpoint showing the node, converting it from the node's AGSKMLViewPoint if possible.
+    private func viewpoint(for node: AGSKMLNode) -> AGSViewpoint? {
```

Comment:
What would you think of making this an asynchronous operation so as not to block the main thread while getting the elevation? Something like this;

```swift
/// Creates an `AGSKMLViewpoint` object based on an `AGSKMLViewpoint`
/// object. The completion handler may be called immediately.
///
/// - Parameters:
///   - kmlViewpoint: A KML viewpoint.
///   - completion: A closure called when the viewpoint has been created.
func makeViewpoint(from kmlViewpoint: AGSKMLViewpoint, completion: (AGSViewpoint?) -> Void) {
    …
}

makeViewpoint(from: kmlViewpoint) { [weak self] (viewpoint) in
    guard let self = self, node == self.node, let viewpoint = viewpoint else { return }
    self.sceneView.setViewpoint(viewpoint)
}
```

Reply:
Good idea. I changed it to execute in the background, though not quite this way.

Reply:
I see what you did. You may not be blocking the main queue, but it looks like you're blocking the global `userInitiated` queue, which may not be the best idea. I assume that getting the elevation at a given point doesn't take an enormous amount of time, but because the queue being blocked is a global queue, who knows what else is being dispatched to it.

Reply:
Are we not supposed to use the globe queues? I can change it to a custom queue, it doesn't matter to me.

Reply:
> Are we not supposed to use the globe queues?

No! Using the global queues is perfectly fine. It is blocking one of them while waiting on work from another queue that I'm concerned about. Tell me if I've got this right:

1. `setSceneViewpoint(for:)` is called on the main queue.
1. `setSceneViewpoint(for:)` dispatches work to the `userInitiated` global queue.
1. `viewpoint(for:)` is called from the `userInitiated` global queue.
1. `sceneSurfaceElevation(for:)` is called from the `userInitiated` global queue.
1. `sceneSurfaceElevation(for:)` waits (blocks) on work dispatched to the `userInteractive` global queue.
1. `AGSSurface.elevation(for:completion:)` is called from the `userInteractive` global queue. The method is asynchronous, so presumably it dispatches work to some arbitrary queue.
1. The `userInteractive` global queue is unblocked when `AGSSurface.elevation(for:completion:)`'s completion handler is called from some arbitrary queue.
1. When `viewpoint(for:)` returns, work is dispatched to the main queue.

Assuming that is correct, a couple observations:

- The dispatch to the `userInteractive` global queue is unnecessary. The only work done on it calls another asynchronous function on another queue.
- The blocking is unnecessary. `viewpoint(for:)` could be made asynchronous with a completion callback. When elevation is needed, it can be retrieved asynchronously and then the viewpoint could be reported to the completion handler on the main thread.

Reply:
Oh, I misunderstood which dispatch you were referring to. I'm going to leave the blocking, for synchronous fetching of the elevation, but remove the extra call to userInteractive. All the viewpoint work will be done on a queue owned by the view controller.